### ISSUE TITLE
React emit should have an '_extends' helper for spread attributes

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#7270]
This is effectively a polyfill for `Object.assign`, and we should consider emitting the specific helper.

**Comment by mhegazy** [comment]
can you elaborate on the request, and why?

**Comment by RyanCavanaugh** [comment]
We currently emit a call to `React.__spread` which has been deprecated for a while now (and might be gone soon?). The official JSX transformer calls `_extends` which is initialized with `Object.assign || (object.assign polyfill)`

**Comment by kuon** [comment]
`__spread` is gone in react 15.0. which means typescript 1.8.9 cannot compile TSX with react 15.0.

**Comment by gaearon** [comment]
If anyone comes here after updating React to 15, a temporary workaround (I think) would be to restore it:

```js
React.__spread = require('object-assign');
```

This would require adding `object-assign` to your dependencies until this is sorted out.

**Comment by gaearon** [comment]
@kuon Thank you for the workaround. 
I would suggest using `object-assign` npm module directly. It is what React uses in 15.

**Comment by kuon** [comment]
Yeah, but as I had lodash in my project, with typescript definition and all, it was easier.

**Comment by gaearon** [comment]
We’ll temporarily bring it back to avoid breaking everyone using TS: https://github.com/facebook/react/pull/6444. However the method is deprecated and will go away in the next major release.

**Comment by DanielRosenwasser** [comment]
Thanks so much @gaearon, we really appreciate the intermittent rollback.

**Comment by kuon** [comment]
Proper workaround without lodash:

In a `lib.d.ts` file, add:

```
declare module 'object-assign' {
    function Assign(target: any, source: any): any;
    module Assign { }
    export = Assign;
}
```

In your main `.ts` file, add:

```
import * as assign from 'object-assign';
(React as any).__spread = assign;
```

**Comment by DanielRosenwasser** [comment]
A fix is out in #7952. Ideally we can get the change into the next nightly.

**Comment by jgillich** [comment]
Just in case someone else wonders, ES6 `Object.assign` works as well:
```
(React as any).__spread = Object.assign;
```
(but you need core-js or another polyfill if you're not targeting ES6).

**Comment by DanielRosenwasser** [comment]
A fix should be out in our nightlies tonight. We'll potentially port the fix to our 1.8 branch.

**Comment by DanielRosenwasser** [comment]
This should be fixed in TypeScript 1.8.10, which is now on npm.

**Comment by basarat** [comment]
Now that we have this for react, how about getting this for normal TypeScript (primarily because libraries out there e.g. `redux` love it) `{...foo,{withSomethingMore}}`

: https://github.com/Microsoft/TypeScript/issues/2103 :rose:
